lib/msprime.c

```diff
@@ -4223,6 +4227,80 @@ msp_add_instantaneous_bottleneck(msp_t *self, double time, int population_id,
     return ret;
 }
 
+/* Add a census event at a specified time (given in generations). */
+static int
+msp_census_event(msp_t *self, demographic_event_t *event)
+{
+    int ret = 0;
+    double time = self->model.model_time_to_generations(&self->model, event->time);
+    avl_tree_t *ancestors;
+    avl_node_t *node;
+    segment_t *seg;
+    int i, j;
+    node_id_t u;
+
+    for (i = 0; i < (int) self->num_populations; i++) {
+        for (j = 0; j < (int) self->num_labels; j++) {
+
+            // Get segment from an ancestor in a population.
+            ancestors = &self->populations[i].ancestors[j];
+            node = ancestors->head;
+
+            while (node != NULL){
+                seg = node->item;
+
+                while (seg != NULL) {
+                    // Add an edge to the edge table.
+                    ret = msp_flush_edges(self);
+                    if (ret != 0) {
+                        goto out;
+                    }
+                    ret = tsk_node_table_add_row(&self->tables->nodes,
+                            MSP_NODE_IS_CEN_EVENT, time, (population_id_t) i, TSK_NULL,
+                            NULL, 0);
+                    u = ret;
+                    // Add an edge joining the segment to the new node.
+                    ret = msp_store_edge(self, seg->left, seg->right, ret, seg->value);
```

Comment:
Need to check the value of ``ret`` here and after node_table_add_row --- an error could have happen during either of these calls and we'd miss them currently.